openjdk-benchmarks/src/jmh/java/org/conscrypt/benchmarks/ServerSocketBenchmark.java

```diff
@@ -0,0 +1,171 @@
+        }
+
+        final SSLServerSocket newServerSocket(String cipher) {
+            try {
+                int port = Util.pickUnusedPort();
+                SSLServerSocket sslSocket =
+                        (SSLServerSocket) serverSocketFactory.createServerSocket(port);
+                sslSocket.setEnabledProtocols(Util.getProtocols());
+                sslSocket.setEnabledCipherSuites(new String[] {cipher});
+                return sslSocket;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    @Param public SslSocketType sslSocketType;
+
+    @Param({"64", "128", "512", "1024", "4096"}) public int messageSize;
+
+    @Param({"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256"}) public String cipher;
+
+    private Client client;
+    private EchoServer server;
+    private byte[] message;
+
+    @Setup
+    public void setup() throws Exception {
+        message = Util.newTextMessage(messageSize);
+        server = new EchoServer(sslSocketType.newServerSocket(cipher), messageSize);
+
+        Future connectedFuture = server.start();
+
+        client = new Client(server.port());
+        client.start();
+
+        // Wait for the initial connection to complete.
+        connectedFuture.get(5, TimeUnit.SECONDS);
+    }
+
+    @TearDown
+    public void teardown() throws Exception {
+        client.stop();
+        server.stop();
+    }
+
+    @Benchmark
+    public void pingPong() throws IOException {
+        client.sendMessage(message);
+        byte[] response = client.readMessage();
+        assertEquals(message.length, response.length);
+    }
+
+    /**
+     * Client-side endpoint. Provides basic services for sending/receiving messages from the client
+     * socket.
+     */
+    private final class Client {
+        private final byte[] buffer = new byte[messageSize];
+        private final SSLSocket socket;
+
+        Client(int port) {
+            try {
+                SSLSocketFactory socketFactory = Util.getJdkSocketFactory();
+                socket = (SSLSocket) socketFactory.createSocket(Util.LOCALHOST, port);
+                socket.setEnabledProtocols(getProtocols());
+                socket.setEnabledCipherSuites(new String[] {cipher});
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        Future<?> start() {
+            try {
+                socket.startHandshake();
+                return Futures.immediateFuture(null);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        void stop() {
+            try {
+                socket.close();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        byte[] readMessage() {
+            try {
+                ByteArrayOutputStream bos = new ByteArrayOutputStream(messageSize);
```

Comment:
Do you need a `ByteArrayOutputStream` here? Why not just pass in an output buffer to write and avoid the `ByteArrayOutputStream`?